src/compiler/checker.ts

```diff
@@ -28417,6 +28417,96 @@ export function createTypeChecker(host: TypeCheckerHost): TypeChecker {
     }
 
     function checkObjectLiteral(node: ObjectLiteralExpression, checkMode?: CheckMode): Type {
+        const contextualType = getContextualType(node, /*contextFlags*/ undefined);
+        const isContextualTypeDependent =
+            node.parent.kind === SyntaxKind.CallExpression &&
+            contextualType &&
+            (contextualType.flags & TypeFlags.TypeParameter) &&
+            contextualType.immediateBaseConstraint &&
+            !!forEachChildRecursively(
+                (contextualType.symbol.declarations![0] as TypeParameterDeclaration).constraint!,
+                node => {
+                    if (
+                        node.kind === SyntaxKind.TypeReference &&
+                        (node as TypeReferenceNode).typeName.kind === SyntaxKind.Identifier &&
+                        ((node as TypeReferenceNode).typeName as Identifier).escapedText ===
+                        (contextualType.symbol.declarations![0] as TypeParameterDeclaration).name.escapedText

```

Comment:
This is probably a brittle way of writing this (unless I'm mistaken) but I don't know any other way that works. I would have used symbols but it seems that they aren't declared at this point.